internal/blockchain/fabric/fabric.go

```diff
@@ -170,61 +176,113 @@ func (f *Fabric) Init(ctx context.Context, config config.Section, callbacks bloc
 	f.capabilities = &blockchain.Capabilities{}
 
 	if fabconnectConf.GetString(ffresty.HTTPConfigURL) == "" {
-		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "url", "blockchain.fabconnect")
+		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "url", "blockchain.fabric.fabconnect")
 	}
+	f.client = ffresty.New(f.ctx, fabconnectConf)
+
 	f.defaultChannel = fabconnectConf.GetString(FabconnectConfigDefaultChannel)
-	f.chaincode = fabconnectConf.GetString(FabconnectConfigChaincode)
-	if f.chaincode == "" {
-		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "chaincode", "blockchain.fabconnect")
-	}
 	// the org identity is guaranteed to be configured by the core
 	f.signer = fabconnectConf.GetString(FabconnectConfigSigner)
 	f.topic = fabconnectConf.GetString(FabconnectConfigTopic)
 	if f.topic == "" {
-		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "topic", "blockchain.fabconnect")
+		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "topic", "blockchain.fabric.fabconnect")
 	}
-
 	f.prefixShort = fabconnectConf.GetString(FabconnectPrefixShort)
 	f.prefixLong = fabconnectConf.GetString(FabconnectPrefixLong)
 
-	f.client = ffresty.New(f.ctx, fabconnectConf)
-
 	wsConfig := wsclient.GenerateConfig(fabconnectConf)
-
 	if wsConfig.WSKeyPath == "" {
 		wsConfig.WSKeyPath = "/ws"
 	}
-
-	f.wsconn, err = wsclient.New(ctx, wsConfig, nil, f.afterConnect)
+	f.wsconn, err = wsclient.New(f.ctx, wsConfig, nil, f.afterConnect)
 	if err != nil {
 		return err
 	}
 
-	f.streams = &streamManager{
-		client: f.client,
-		signer: f.signer,
-	}
-	batchSize := fabconnectConf.GetUint(FabconnectConfigBatchSize)
-	batchTimeout := uint(fabconnectConf.GetDuration(FabconnectConfigBatchTimeout).Milliseconds())
+	f.streams = &streamManager{client: f.client, signer: f.signer}
+	batchSize := f.fabconnectConf.GetUint(FabconnectConfigBatchSize)
+	batchTimeout := uint(f.fabconnectConf.GetDuration(FabconnectConfigBatchTimeout).Milliseconds())
 	if f.initInfo.stream, err = f.streams.ensureEventStream(f.ctx, f.topic, batchSize, batchTimeout); err != nil {
 		return err
 	}
 	log.L(f.ctx).Infof("Event stream: %s", f.initInfo.stream.ID)
-	location := &Location{
-		Channel:   f.defaultChannel,
-		Chaincode: f.chaincode,
-	}
-	if f.initInfo.sub, err = f.streams.ensureSubscription(f.ctx, location, f.initInfo.stream.ID, batchPinEvent); err != nil {
-		return err
-	}
 
 	f.closed = make(chan struct{})
 	go f.eventLoop()
 
 	return nil
 }
 
-func (f *Fabric) Start() error {
+func (f *Fabric) resolveFireFlyContract(ctx context.Context, contractIndex int) (chaincode, fromBlock string, err error) {
+
+	if f.contractConfSize > 0 || contractIndex > 0 {
+		// New config (array of objects under "fireflyContract")
+		if contractIndex >= f.contractConfSize {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgInvalidFireFlyContractIndex, fmt.Sprintf("blockchain.fabric.fireflyContract[%d]", contractIndex))
```

Comment:
Might be worth talking to @jimthematrix about the Fabric side of this, and whether in practice we use this feature you've added for jumping chaincode names, vs. just upgrading the Fabric chaincode in-place. Not a blocker to this merging, but something to think about in the next step when developing the v.next contract code, and testing this journey end to end with a Fabric environment.